packages/abi-coder/src/fragments/param-type.ts

```diff
@@ -0,0 +1,175 @@
+import { Logger } from '@ethersproject/logger';
+import { defineReadOnly } from '@ethersproject/properties';
+
+import { arrayRegEx, enumRegEx, structRegEx } from '../abi-coder';
+
+const version = 'abi/5.6.4';
+export const abiLogger = new Logger(version);
+
+export interface JsonFragmentType {
+  readonly name?: string;
+  readonly indexed?: boolean;
+  readonly type: string;
+  readonly components?: ReadonlyArray<JsonFragmentType>;
+  readonly typeArguments?: ReadonlyArray<JsonFragmentType>;
+}
+
+const CONSTRUCTOR_GUARD = {};
+
+function populate(object: ParamType, params: { type?: string } & ParamTypeProps) {
+  Object.keys(params).forEach((key) => {
+    const paramTypeKey = key as keyof ParamTypeProps;
+    const value = params[paramTypeKey];
+
+    defineReadOnly(object, paramTypeKey, value);
+  });
+}
+
+const paramTypeArray = /^(.*)\[([0-9]*)\]$/;
+
+export interface ParamTypeProps {
+  // The local name of the parameter (of null if unbound)
+  readonly name?: string;
+
+  // The fully qualified type (e.g. "address", "tuple(address)", "uint256[3][]"
+  readonly type: string;
+
+  // The base type (e.g. "address", "tuple", "array")
+  readonly baseType?: string;
+
+  // Indexable Paramters ONLY (otherwise null)
+  readonly indexed?: boolean;
+
+  // Tuples ONLY: (otherwise null)
+  //  - sub-components
+  readonly components?: Array<ParamType>;
+
+  // Tuples ONLY: (otherwise null)
+  //  - sub-components
+  readonly typeArguments?: Array<ParamType>;
+
+  // Arrays ONLY: (otherwise null)
+  //  - length of the array (-1 for dynamic length)
+  //  - child type
+  readonly arrayLength?: number;
+  readonly arrayChildren?: ParamType;
+
+  readonly isParamType?: boolean;
+}
+
+export class ParamType implements ParamTypeProps {
+  readonly name?: string;
+  readonly type!: string;
+  readonly baseType?: string;
+  readonly indexed?: boolean;
+  readonly components?: Array<ParamType>;
+  readonly typeArguments?: Array<ParamType>;
+  readonly arrayLength?: number;
+  readonly arrayChildren?: ParamType;
+  readonly isParamType?: boolean;
+
+  constructor(constructorGuard: typeof CONSTRUCTOR_GUARD, params: ParamTypeProps) {
+    if (constructorGuard !== CONSTRUCTOR_GUARD) {
+      abiLogger.throwError('use fromString', Logger.errors.UNSUPPORTED_OPERATION, {
+        operation: 'new ParamType()',
+      });
+    }
+    populate(this, params);
+
+    const match = this.type.match(paramTypeArray);
+    if (match && match[1]) {
```

Comment:
Maybe invert the if and inside the else use destruct on the match to add more sense?

```ts
const [type, size] = match;
```

Reply:
as mentioned in prev comment, refactored this to use `stringRegEx` instead